File: pico/Sense_HAT_C_Demo/STM32/SHTC3/Core/Src/SHTC3.c
#include <math.h>
#include "SHTC3.h"

char SHTC3_CheckCrc(uint8_t data[],unsigned char len,unsigned char checksum)
{
  unsigned char bit;        // bit mask
  unsigned char crc = 0xFF; // calculated checksum
  unsigned char byteCtr;    // byte counter
  // calculates 8-Bit checksum with given polynomial
  for(byteCtr = 0; byteCtr < len; byteCtr++) {
    crc ^= (data[byteCtr]);
    for(bit = 8; bit > 0; --bit) {
      if(crc & 0x80) {
        crc = (crc << 1) ^ CRC_POLYNOMIAL;
      } else {
        crc = (crc << 1);
      }
    }
  }

  // verify checksum
  if(crc != checksum) {                 
    return 1;                       //Error
  } else {
    return 0;                       //No error
  }       
}
void SHTC3_WriteCommand(unsigned short cmd)
{   
    uint8_t buf[] = {(cmd>>8) ,cmd};
    HAL_I2C_Master_Transmit(&hi2c1,SHTC3_I2C_ADDRESS,buf,2,100);          
                                                 //1:error 0:No error
}
void SHTC3_WAKEUP(void)
{     
    SHTC3_WriteCommand(SHTC3_WakeUp);                  // write wake_up command  
    delay_us(300);                          //Delay 300us
      
}
void SHTC3_SLEEP(void)
{    
 //   bcm2835_i2c_begin();
    SHTC3_WriteCommand(SHTC3_Sleep);                        // Write sleep command
      
}

void SHTC_SOFT_RESET(void)
{   
    SHTC3_WriteCommand(SHTC3_Software_RES);                 // Write reset command
    delay_us(300);                                 //Delay 300us
     
}


float TH_Value()
{
		unsigned short TH_DATA;
		uint8_t checksum;
    uint8_t buf[3];
		SHTC3_WriteCommand(SHTC3_NM_CD_ReadTH);                 //Read temperature first,clock streching disabled (polling)
    HAL_Delay(20);
    HAL_I2C_Master_Receive(&hi2c1,SHTC3_I2C_ADDRESS,buf,3,1000);

   checksum=buf[2];
   if(!SHTC3_CheckCrc(buf,2,checksum))
        TH_DATA=(buf[0]<<8|buf[1]);
	 return (175 * (float)TH_DATA / 65536.0f - 45.0f);       //Calculate temperature value
}

float RH_Value()
{
		unsigned short RH_DATA;
		uint8_t checksum;
		uint8_t buf[3];
		SHTC3_WriteCommand(SHTC3_NM_CD_ReadRH);                 //Read temperature first,clock streching disabled (polling)
		HAL_Delay(20);
		HAL_I2C_Master_Receive(&hi2c1,SHTC3_I2C_ADDRESS,buf,3,1000);

    checksum=buf[2];
    if(!SHTC3_CheckCrc(buf,2,checksum))
        RH_DATA=(buf[0]<<8|buf[1]);
		return (100 * (float)RH_DATA / 65536.0f);									//Calculate temperature value
}

File: pico/Sense_HAT_C_Pi/RaspberryPi/SHTC3/python/SHTC3.py
#!/usr/bin/python
# -*- coding:utf-8 -*-
import time
import lgpio as sbc

SHTC3_I2C_ADDRESS   = 0x70

SHTC3_ID            = 0xEFC8
CRC_POLYNOMIAL      = 0x0131
SHTC3_WakeUp        = 0x3517
SHTC3_Sleep         = 0xB098
SHTC3_Software_RES  = 0x805D
SHTC3_NM_CD_ReadTH  = 0x7866
SHTC3_NM_CD_ReadRH  = 0x58E0

class SHTC3():
    def __init__(self,sbc,bus,address,flags = 0):
        self._sbc = sbc
        self._fd = self._sbc.i2c_open(bus, address, flags)
        self.SHTC_SOFT_RESET()

    def SHTC3_CheckCrc(self,data,len,checksum):
        crc = 0xFF
        for byteCtr in range(0,len):
            crc = crc ^ data[byteCtr]
            for bit in range(0,8):
                if crc & 0x80:
                    crc = (crc << 1) ^ CRC_POLYNOMIAL
                else:
                    crc = crc << 1
        if crc == checksum:
            return True
        else:
            return False
    
    def SHTC3_WriteCommand(self,cmd):
        self._sbc.i2c_write_byte_data(self._fd,cmd >> 8,cmd & 0xFF)
    
    def SHTC3_WAKEUP(self):
        self.SHTC3_WriteCommand(SHTC3_WakeUp) # write wake_up command
        time.sleep(0.01) # Prevent the system from crashing 
    
    def SHTC3_SLEEP(self):
        self.SHTC3_WriteCommand(SHTC3_Sleep) # Write sleep command
        time.sleep(0.01)
    
    def SHTC_SOFT_RESET(self):
        self.SHTC3_WriteCommand(SHTC3_Software_RES) # Write reset command
        time.sleep(0.01)
    
    def SHTC3_Read_TH(self): # Read temperature 
        self.SHTC3_WAKEUP()
        self.SHTC3_WriteCommand(SHTC3_NM_CD_ReadTH)
        time.sleep(0.02)
        (count,buf) = self._sbc.i2c_read_device(self._fd,3)
        if self.SHTC3_CheckCrc(buf,2,buf[2]):
            return (buf[0]<<8|buf[1]) * 175 / 65536 - 45.0 # Calculate temperature value
        else:
            return 0 # Error
    
    def SHTC3_Read_RH(self): # Read humidity 
        self.SHTC3_WAKEUP()
        self.SHTC3_WriteCommand(SHTC3_NM_CD_ReadRH)
        time.sleep(0.02)
        (count,buf) = self._sbc.i2c_read_device(self._fd,3)
        if self.SHTC3_CheckCrc(buf,2,buf[2]) :
            return 100 * (buf[0]<<8|buf[1]) / 65536 # Calculate humidity value
        else:
            return 0  # Error

if __name__ == "__main__":
    try:
        shtc3 = SHTC3(sbc, 1, SHTC3_I2C_ADDRESS)
        
        while True:
            print("Temperature = %6.2f°C , Humidity = %6.2f%%"%(shtc3.SHTC3_Read_TH(),shtc3.SHTC3_Read_RH()))
    except:
        print ("\nProgram end")
        exit()